Add manual PID tuning with TorqueCurrentFOC page (#161)
CrossTheRoadElec/Phoenix6-Documentation: source/docs/application-notes/manual-pid-tuning.rst

   From here, I set a setpoint of 0.1 and dial in kS to just before it starts moving. I increase it to 1, 2, and 4 when it starts moving. From here, I dial it down to 3 where it doesn't move, and back up to 3.5, 3.7 where it moves again. I check 3.6 and see it doesn't move, so I leave kS at 3.6 amps.

   Now it's time for kP/kD tuning. I bring kP up to 1, 2, 4, 8, 16, and 32 before I get significant overshoot, where I dial kD in to 1, 2, 4, and 8 before that overshoot is gone. kP keeps increasing to 64 and 128, then kD goes up to 16 and 32 before it's back to kP. I go up to 256 and 512 where I notice a bit of oscillation, and I may be near the limit at this point. kD increases to 64 and I get oscillation on the way to the target, so I bring it down to 50 then 40 before I'm happy with it. There's still a little oscillation at the target, but it's minimal.

   I check with other setpoints of -0.1, 0.4, 0.6 and confirm the movement looks good, and say the PID tuning is done.

Profiled Tuning
^^^^^^^^^^^^^^^

> Profiled tuning can be treated much the same way as normal tuning, but the introduction of a profile the system should follow means much of the response can be calculated in advance with feed-forwards.

Reviewer: ```suggestion
Profiled tuning follows a similar process as normal tuning, but with the introduction of a motion profile, this means much of the response can be calculated in advance with feed-forwards.
```

You don't have to use my recommendation, but the sentence is stiff. The latter half is particularly stiff.

Similar math can be done for every other gain constant to find a good starting point.

Specific Response Tuning
------------------------

These general guidelines are great for understanding what's happening in a closed loop controller and how to forward-calculate what a reasonable starting point is. However, the specific mechanism you're tuning is going to affect how to find the ideal gains and what gains you should be using in the first place.

Flywheel Tuning with TorqueCurrentFOC
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Below is a list of steps and a simulator that provides the opportunity to try tuning a flywheel system using PID with TorqueControl. The Red line is the setpoint of the flywheel controller, purple is the current velocity, and the green line is the current of the motor in stator-amps.

This particular flywheel has a maximum velocity of 100 rps.

Tuning a flywheel is largely done with the following steps:
 1. Zero all PID gains.
 2. Set a high setpoint (typically 8/10th the maximum velocity).
 3. Increase kS until the wheel starts moving, then back off to just before that movement.

>  4. Set kP to a very low number (typically 10 / setpoint is a good starting point).

Reviewer 2: After some further testing, the starting kP you want is going to be somewhere around 0.5 to 1 * `RotorToSensorRatio` * `SensorToMechanismRatio`. The documentation has been updated to match.

   So I double kP again to 128, then to 256 where I notice it oscillates a bit. I try to stop this oscillation by increasing kD to 16, then to 32 where I notice it's always oscillating. This means I've reached the limit of the system, and need to back off on gains a bit.

   I reduce kD back to 16 where I notice a bit of oscillation on its way to the setpoint, and start dialing back kP. I start with a kP of 200, where it's overdamped and oscillating on its way to the setpoint. So I reduce kD to 12.

   From here I continue to reduce kP to 180, then 150 where I notice the oscillation on its way to the setpoint again. Reduce kD again to 10, and decrease kP to 140, then 130 where I see oscillation on its way to setpoint again.

   Reduce kD even more to 9, and the system response looks relatively good at this point. Now it's time to play with different setpoint. Any setpoint within 1 rotation looks good, which is appropriate for a turret. However, let's say I'm not tuning a turret anymore, but some other position controller where a setpoint of, say, 20 is appropriate. When I set a setpoint of 20, I notice significant overshoot that I should correct in PID.

>    At this point, I know that my kD can't go much higher otherwise I have oscillation on my way to the setpoint at smaller setpoints. So I try to stop the oscillation only with kP. Reducing it to 120, 110, 100, then finally 90 before the overshoot stops. I check back with my 0.1 setpoint to make sure it's still good, but now it looks overdamped.

Reviewer 2: This entire section has been redone based on changes to the underlying simulation (the system is now on a 20:1 gear reduction). The new documented gains shouldn't see any overshoot at least in the +/- 0.5 rot case. As always, without a motion profile, your gains will vary based on what travel distance you prioritize.

   3. The highest possible kG that prevents the system from moving is the upper bound of the gravity and friction component.
   4. Setting kG to the middle point of the lower and upper bounds is a good approximation for the true effect of gravity, removing the force of friction.
   5. A nearby setpoint ensures the system response should be relatively small to start with when tuning.
   6. The kS gain is meant to reduce the effect of friction, so the largest possible value that still prevents the system from moving will reduce the effect of friction in general.
   7. The kP gain will control how quickly the system gets to the setpoint, however in TorqueCurrentFOC modes there is no natural dampening force, so overshoot is expected at the beginning. Once that happens kD should be tuned.
   8. The kD gain will effectively slow down the system as it reaches the setpoint, increasing it will increase the force slowing it down, so it should be increased until the system no longer overshoots.

>    9. In general, the system wants as high a kP gain as possible to decrease the time taken to get to the setpoint. This also requires a high kD gain to properly dampen the system. The limit of how high the kP/kD term can be is determined by the system latency, at which point the oscillation is impossible to avoid. The goal of repeating steps 7 and 8 is to find that limit.

Reviewer 3: I think this explanation that the goal is to have as high a kP gain as possible, limited by oscillation, is very important. Perhaps this should be stated above outside of the mechanism specific sections or at least not "hidden" in the "why" section.

   I check with other setpoints of -0.1, 0.4, 0.6 and confirm the movement looks good, and say the PID tuning is done.

Profiled Tuning
^^^^^^^^^^^^^^^
Profiled tuning can be treated much the same way as tuning a normal PID, but the introduction of a profile means much of the response can be calculated in advance with feed-forwards. This results in most of the work being done due to feed forward, and the feedback gains being used to account for any error in the system.

In Phoenix 6, you can either generate your own profile and feed in the position and velocity setpoints, or use MotionMagic® and let the Talon generate the profile for you. In either case the Talon will have Velocity and/or Acceleration setpoints that it can use the kV and kA feedforward terms on, for more accurate profile following.

> The example below uses a pre-generated profile for the system to follow, and the general steps to tune it are below:

Reviewer 3: I would recommend stating what each color in the graph represents like was done for the first two examples.

> Red is the position setpoint in rotations, purple is the current position, blue is the velocity setpoint in rotations/sec, brown is the current velocity, yellow is the acceleration setpoint in rotations/sec/sec, orange is the current acceleration, and green is the stator current in amps

1. Gains are canonical

   - :math:`k_{P} = \frac{\mathrm{motor\_output}}{\mathrm{error}}`
      - The amount of output to apply per unit of error in the system.

   - :math:`k_{I} = \frac{\mathrm{motor\_output}}{\mathrm{error} \cdot \mathrm{sec}}`
      - The amount of output to apply per unit of error for every second of that error.

   - :math:`k_{D} = \frac{\mathrm{motor\_output}}{\frac{\mathrm{error}}{\mathrm{sec}}}=\frac{\mathrm{motor\_output} \cdot \mathrm{sec}}{\mathrm{error}}`
      - The amount of output to apply per change in error over time.

   - :math:`k_{S} = \mathrm{motor\_output}`

>       - A static amount of output to apply typically used to overcome friction.

Reviewer: > - A static amount of output to apply typically used to overcome friction.

to

> - A static or constant amount of output to apply. This is typically used to overcome friction.

   - :math:`k_{G} = \frac{\mathrm{motor\_output}}{\cos(\mathrm{angle})}` if mechanism is arm; :math:`k_{G} = \mathrm{motor\_output}` if mechanism is an elevator
      - The amount of output to apply to counteract the force of gravity.

   - :math:`k_{V} = \frac{\mathrm{motor\_output}}{\mathrm{vel}}`

>       - The amount of output to apply per target velocity. In Voltage modes this is a feed forward to counteract the back-emf of the motor, in Current modes this is a feed forward to counteract the force of drag on the system.

Reviewer: > The amount of output to apply per target velocity. In Voltage modes this is a feed forward to counteract the back-emf of the motor, in Current modes this is a feed forward to counteract the force of drag on the system.

to

> The amount of output to apply per target velocity. In Voltage control modes, this is a feed forward to counteract the back-emf of the motor. In Current modes, this is a feed forward to counteract the force of drag on the system.

Also add cross links

 10. Increase kP until the flywheel oscillates, then back off to just before that oscillation.
 11. Verify gains hold for expected velocities.

.. dropdown:: "Why" for each step

   1. We start with PID gains at 0 to isolate as many of the forces in play as possible, and iteratively get closer to the "ideal" gains.
   2. There needs to be a setpoint for the parameters to take affect, and a higher setpoint sets up the following steps.
   3. The kS gain is meant to reduce the effect of friction without it moving the mechanism on its own, so getting it as close to breaking friction without it actually breaking friction is ideal. However, this step alone only accounts for static friction which isn't ideal in a flywheel, where it'll typically be experiencing rolling friction. This is managed in a later step.

>    4. A low kP will exacerbate the requirement for a good kS and kV. If this gain is too high it'll mask a "bad" kS and kV during the kS/kV tuning.

Reviewer: > A low kP will exacerbate the requirement for a good kS and kV. If this gain is too high it'll mask a "bad" kS and kV during the kS/kV tuning.

to

> A low kP will magnify the requirement for a good kS and kV. If this gain is too high it'll mask a "bad" kS and kV during the kS/kV tuning.

The following steps cover the general idea:
 1. Zero all PID gains.
 2. Set a setpoint relatively nearby (typically 0.1 mechanism rotations).
 3. Increase kS until the turret starts moving, then back off to just before that movement.
 4. Increase kP until you notice significant overshoot.
 5. Increase kD until the overshoot stops happening.
 6. Repeat steps 4 and 5 until increasing kD results in more oscillation, or until the system oscillates on its way to the setpoint. If oscillation on the way to setpoint is seen, decrease kD until it stops. If overshoot in general is happening and kD is already at max, reduce kP until it stops.
 7. Verify gains work for other setpoints as well. Tune kP/kD as appropriate for most general cases.

.. note:: Values of kP=200, kD=15 demonstrate the "oscillates on its way to the setpoint" case for setpoints within 1 rotation.

.. dropdown:: "Why" for each step

   1. We start with PID gains at 0 to isolate as many of the forces in play as possible, and iteratively get closer to the "ideal" gains.

>    2. A close-by setpoint ensures the system response should be relatively small to start with when tuning.

Reviewer: >    2. A close-by setpoint ensures the system response should be relatively small to start with when tuning.

to

>    2. A nearby setpoint ensures the system response should be relatively small to start with when tuning.

 6. Increase kS until the arm starts moving, then back off to just before that movement.
 7. Increase kP until you notice significant overshoot.
 8. Increase kD until the overshoot stops happening.
 9. Repeat steps 7 and 8 until increasing kD results in more oscillation, or until the system oscillates on its way to the setpoint. If oscillation on the way to setpoint is seen, decrease kD until it stops. If overshoot in general is happening and kD is already at max, reduce kP until it stops.
 10. Verify gains work for other setpoints as well. Tune kP/kD as appropriate for most general cases.

.. dropdown:: "Why" for each step

   1. We start with PID gains at 0 to isolate as many of the forces in play as possible, and iteratively get closer to the "ideal" gains.

>    2. The kG gain is meant to counteract the force of gravity, however the force of friction is also at play in an arm. The lowest possible kG that prevents the system from moving finds the lower bound of the gravity and friction component.

Reviewer: > The kG gain is meant to counteract the force of gravity, however the force of friction is also at play in an arm. The lowest possible kG that prevents the system from moving finds the lower bound of the gravity and friction component.

to

> The kG gain is meant to counteract the force of gravity, however the force of friction is also at play in an arm. The lowest possible kG that prevents the system from moving is the lower bound of the gravity and friction component.

>    3. The highest possible kG that prevents the system from moving finds the upper bound of the gravity and friction component.

Reviewer: > 3. The highest possible kG that prevents the system from moving finds the upper bound of the gravity and friction component.

to 

> 3. The highest possible kG that prevents the system from moving is the upper bound of the gravity and friction component.

>    4. Setting kG to the middle point of the lower and upper bound is a good approximation for the true effect of gravity minus the force of friction.

Reviewer: 4. Setting kG to the middle point of the lower and upper bound is a good approximation for the true effect of gravity minus the force of friction.

to

4. Setting kG to the middle point of the lower and upper bounds is a good approximation for the true effect of gravity minus the force of friction.

>    5. A close-by setpoint ensures the system response should be relatively small to start with when tuning.

Reviewer: > 5. A close-by setpoint ensures the system response should be relatively small to start with when tuning.

to

> 5. A nearby setpoint ensures the system response should be relatively small to start with when tuning.
